tools/data_gen/src/main.rs

```diff
@@ -0,0 +1,173 @@
+            vec!["D0_SIZE", "D1_SIZE", "D2_SIZE", "D3_SIZE"],
+        );
+        m
+    };
+}
+
+struct CellData {
+    name: String,
+    width: u64,
+    sizes: Vec<u64>,
+}
+
+enum SizeVec {
+    Quad(Vec<Vec<Vec<Vec<u64>>>>),
+    Triple(Vec<Vec<Vec<u64>>>),
+    Double(Vec<Vec<u64>>),
+    Single(Vec<u64>),
+}
+
+impl Serialize for SizeVec {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        match &*self {
+            SizeVec::Single(v) => {
+                serializer.serialize_newtype_variant("SizeVec", 3, "data", &v)
+            }
+            SizeVec::Double(v) => {
+                serializer.serialize_newtype_variant("SizeVec", 2, "data", &v)
+            }
+            SizeVec::Triple(v) => {
+                serializer.serialize_newtype_variant("SizeVec", 1, "data", &v)
+            }
+            SizeVec::Quad(v) => {
+                serializer.serialize_newtype_variant("SizeVec", 0, "data", &v)
+            }
+        }
+    }
+}
+
+#[derive(FromArgs)]
+/// Used to read a file
+struct FilePath {
+    /// file path to read data from
+    #[argh(positional, from_str_fn(read_path))]
+    file: Option<PathBuf>,
+}
+
+fn read_path(path: &str) -> Result<PathBuf, String> {
+    Ok(Path::new(path).into())
+}
+
+fn main() -> CalyxResult<()> {
+    let fp: FilePath = argh::from_env();
+
+    let lib_path: &Path = &Path::new("../../../calyx");
+
+    let ws = frontend::Workspace::construct(&fp.file, lib_path)?;
+    let ctx: ir::Context = (ir::from_ast::ast_to_ir(ws))?;
+
+    let comp = ctx
+        .components
+        .into_iter()
+        .find(|comp| comp.name == ctx.entrypoint)
+        .expect("No top-level component found.");
+
+    let data_vec: Vec<CellData> = comp
+        .cells
+        .iter()
+        .filter_map(|cell| get_data(cell))
+        .collect();
+
+    let mut map = Map::new();
+
+    for CellData {
+        name,
+        width,
+        mut sizes,
+    } in data_vec
+    {
+        let json_comp: serde_json::Value = gen_comp(&mut sizes, width);
+        map.insert(name, json_comp);
+    }
+
+    let json_map: Value = map.into();
+    println!("{}", json_map.to_string());
+    Ok(())
+}
+
+//generates a json value associated with sizes_vec and width
+fn gen_comp(sizes_vec: &mut Vec<u64>, width: u64) -> serde_json::Value {
+    let initial: SizeVec =
+        SizeVec::Single(vec![0; sizes_vec.pop().unwrap().try_into().unwrap()]);
+    sizes_vec.reverse();
+    let data_vec: SizeVec =
+        sizes_vec.iter().fold(initial, |acc, x| accumulate(*x, acc));
+    let data = serde_json::to_value(data_vec).unwrap();
+    json!({
+        "data": data.get("data").unwrap(),
+        "format": {
+            "numeric_type": "bitnum",
+            "is_signed": false,
+            "width": width,
+        }
+    })
+}
+
+//creates a vector consisting of i v's.
+fn add_dimension<T: std::clone::Clone>(i: u64, v: Vec<T>) -> Vec<Vec<T>> {
```

Comment:
Why does this have the type `(u64, Vec<T>) -> Vec<Vec<T>>` instead of `(u64, T) -> Vec<T>`?